### PR TITLE
Option to enforce authority check (`mine=True`)

### PR DESCRIPTION
**Pull request by Leobouloc**
## Purpose

Currently, the API has no restriction to the data / endpoints available to users. All authenticated users have unlimited read and write access, through all endpoints. The purpose of this PR is to add an option to restrict data reading to "own" statements, thus enabling multi tenancy.

## Proposal

It is proposed to add an environment variable named `RALPH_LRS_RESTRICT_BY_AUTHORITY` (default `False`) which, if set to `True` , will IMPLICITLY enforce the use of the query parameter `mine=True` when fetching statements. This in effect forces all users to query only their own statements.

NB: this option relies on `authority` queries being properly implemented for each backend.

NB 2: in a future implementation, "scopes" and/or "roles" may be added. For "admin" type users or users with "all/read" scopes, the mandatory use of `mine` should be removed.

- [x] implicitly enforce `mine=True` when `RUNSERVER_RESTRICT_BY_AUTHORITY==True`
- [x] add test 
- [x] check / add tests to all backends to ensure that authority queries are properly implemented. NB: tested against all backends in `test_api_statements_get_statements_mine`
- [x] test that authority applies even when fetching by ID

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ have an authority field matching that of the user
 - Backends: add `max_statements` option to `AsyncLRSHTTP`
 - API: Incoming statements are enriched with `id`, `timestamp`, `stored`
   and `authority`
+- API: Add `TODO: PUT NAME HERE WHEN DEFINITIVE`option making `?mine=True` mandatory
```
Comment (Leobouloc) [inline CHANGELOG.md:32]: do we agree with `RALPH_RUNSERVER_RESTRICT_BY_AUTHORITY` ?

**Comment by Leobouloc** [comment]
if authority is left to a user to create. A user may choose the same authority as another and thus see all their statements. Should we leave this option open (the admin has responsibility to check), or disable duplicate authorities ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,6 +29,8 @@ have an authority field matching that of the user
 - Backends: add `max_statements` option to `AsyncLRSHTTP`
 - API: Incoming statements are enriched with `id`, `timestamp`, `stored`
   and `authority`
+- API: Add `TODO: PUT NAME HERE WHEN DEFINITIVE`option making `?mine=True`
```
Comment (Leobouloc) [inline CHANGELOG.md:32]: don't forget this

File: CHANGELOG.md (hunk 1 above)
Comment (Leobouloc) [reply CHANGELOG.md]: It was decided to use `RALPH_LRS_RESTRICT_BY_AUTHORITY`

File: docs/api.md
Comment (quitterie-lcs) [inline docs/api.md:1]: Can you update line 138 as follows ? There is a small typo, that makes reviewing confusing in markdown

```
< HTTP/1.1 401 Unauthorized
```

File: src/ralph/api/auth/user.py [hunk 3]
```diff
@@ -1,10 +1,24 @@
 """Authenticated user for the Ralph API."""
 
-from typing import Dict, List
+from enum import Enum
+from typing import Dict, List, Literal
 
 from pydantic import BaseModel
 
 
+Scope = Literal[
```
Comment (quitterie-lcs) [inline src/ralph/api/auth/user.py:9]: Maybe integrate the literal directly in the Pydantic model

Comment (Leobouloc) [reply src/ralph/api/auth/user.py:9]: I left it out for future use (when actually managing scopes)

**Review by wilbrdt — APPROVED** [review]
LGTM! :tada:

File: src/ralph/conf.py [hunk 4]
```diff
@@ -376,11 +376,13 @@ class AuthBackends(Enum):
     RUNSERVER_AUTH_BACKEND: AuthBackends = AuthBackends.BASIC
     RUNSERVER_AUTH_OIDC_AUDIENCE: str = None
     RUNSERVER_AUTH_OIDC_ISSUER_URI: AnyHttpUrl = None
-    RUNSERVER_BACKEND: Literal["clickhouse", "es", "mongo"] = "es"
+    RUNSERVER_BACKEND: Literal["es", "mongo", "clickhouse"] = "es"
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/conf.py:379]: hmmm...

File: tests/helpers.py [hunk 5]
```diff
@@ -48,3 +49,72 @@ def _all_but_statements(response):
         assert statements_are_equivalent(
             statement_1, statement_2
         ), "Statements in get responses are not equivalent, or not in the same order."
+
+
+def create_mock_activity(id_: int = 0):
+    """Create distinct activites with valid IRIs.
+
+    args:
```
Comment (SergioSim) [inline tests/helpers.py:57]:
```suggestion
    Args:
```

File: tests/helpers.py [hunk 6]
```diff
@@ -48,3 +49,72 @@ def _all_but_statements(response):
...
+    """Create distinct mock agents with a given Inverse Functional Identifier.
+
+    args:
+        ifi: Inverse Functional Identifier. Possible values are:
+            'mbox', 'mbox_sha1sum', 'openid' and 'account'.
+        id_: An integer used to uniquely identify the created agent.
+            If ifi=="account", agent equality requires same (id_, home_page_id)
+        home_page (optional): The value of homePage, if ifi=="account"
+        name (optional): name of the agent (NB: do not confuse with account.name
+            with ifi=="account", or "username", as used in credentials)
```
Comment (SergioSim) [inline tests/helpers.py:83]: Missing `use_object_type` ?

File: src/ralph/api/routers/statements.py [hunk 7]
```diff
@@ -326,6 +326,15 @@ async def get(
             json.loads(query_params["agent"])
         )
 
+    if not mine:
+        if settings.LRS_RESTRICT_BY_AUTHORITY:
```
Comment (SergioSim) [inline src/ralph/api/routers/statements.py:330]:
```suggestion
    if settings.LRS_RESTRICT_BY_AUTHORITY:
```
Could this work too (with less nesting)?

**Review by SergioSim — APPROVED** [review]
LGTM) :+1: 
image

File: tests/helpers.py [hunk 8]
```diff
@@ -48,3 +49,72 @@ def _all_but_statements(response):
...
+def create_mock_agent(
+    ifi: str,
+    id_: int,
+    home_page_id=None,
+    name: str = None,
+    use_object_type: bool = True,
+):
+    """Create distinct mock agents with a given Inverse Functional Identifier.
+
+    args:
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply tests/helpers.py:76]: Yes

File: tests/helpers.py (hunk 5 above)
Comment (Leobouloc) [reply tests/helpers.py:57]: Yes

File: src/ralph/api/routers/statements.py (hunk 7 above)
Comment (Leobouloc) [reply src/ralph/api/routers/statements.py:330]: Good point ! Changed this.

File: tests/helpers.py (hunk 6 above)
Comment (Leobouloc) [reply tests/helpers.py:83]: Added it.

File: CHANGELOG.md [hunk 9]
```diff
@@ -38,6 +38,8 @@ have an authority field matching that of the user
   and `authority`
 - Backends: update `statementId` and `voidedStatementId` to snake_case,
   with camelCase alias, in `LRSStatementsQuery`
+- API: Add `RALPH_LRS_RESTRICT_BY_AUTHORITY`option making `?mine=True`
```
Comment (quitterie-lcs) [inline CHANGELOG.md:41]:
```suggestion
- API: Add `RALPH_LRS_RESTRICT_BY_AUTHORITY` option making `?mine=True` 
```

**Review by quitterie-lcs — APPROVED** [review]

**Pull request merged**